source/s3_client.c

```diff
@@ -2054,3 +2054,159 @@ struct aws_byte_cursor aws_s3_meta_request_resume_token_upload_id(
+ *  3.1 Adjust the current timeout against the expected timeout, via 0.99 * <current timeout> + 0.01 * <expected
+ *      timeout> to get closer to the expected timeout.
+ * 4. Once request finishes with timeout. We check the timeout rate.
+ *  4.1 If timeout rate is larger than 0.1%, we increase the timeout by 100ms (Check the timeout when the
+ *      request was made, if the updated timeout is larger than the expected, skip update).
+ *  4.2 If timeout rate is larger than 1%, we increase the timeout by 1 secs (If needed). And clear the rate
+ *      to get the exact rate with new timeout.
+ *  4.3 Once the timeout is larger than 5 secs, we stop the process.
+ *
+ * Invoked from `s_s3_auto_ranged_put_send_request_finish`.
+ */
+void aws_s3_client_update_upload_part_timeout(
+    struct aws_s3_client *client,
+    struct aws_s3_request *finished_upload_part_request,
+    int finished_error_code) {
+
+    aws_s3_client_lock_synced_data(client);
+    struct aws_s3_upload_part_timeout_stats *stats = &client->synced_data.upload_part_stats;
+    if (stats->stop_timeout) {
+        /* Timeout was disabled */
+        goto unlock;
+    }
+
+    struct aws_s3_request_metrics *metrics = finished_upload_part_request->send_data.metrics;
+    size_t current_timeout_ms = aws_atomic_load_int(&client->upload_timeout_ms);
+    uint64_t current_timeout_ns =
+        aws_timestamp_convert(current_timeout_ms, AWS_TIMESTAMP_MILLIS, AWS_TIMESTAMP_NANOS, NULL);
+    uint64_t updated_timeout_ns = 0;
+
+    switch (finished_error_code) {
+        case AWS_ERROR_SUCCESS:
+            /* We only interested in request succeed */
+            ++stats->num_upload_requests_succeed;
+            if (stats->num_upload_requests_succeed <= 10) {
+                /* Gether the data */
+                uint64_t request_time_ns =
+                    metrics->time_metrics.receive_end_timestamp_ns - metrics->time_metrics.send_start_timestamp_ns;
+                stats->request_time_sum_ns += request_time_ns;
+                ++stats->number_request_time;
+                if (stats->num_upload_requests_succeed == 10) {
+                    /* Decide we need a timeout or not */
+                    uint64_t average_request_time_ns = stats->request_time_sum_ns / stats->number_request_time;
+                    if (average_request_time_ns >= s_upload_timeout_threshold_ns) {
+                        /* We don't need a timeout, as retry will be slower than just wait for the server to response */
+                        stats->stop_timeout = true;
+                    } else {
+                        /* Start the timeout by 1 secs */
+                        aws_atomic_store_int(&client->upload_timeout_ms, 1000);
+                    }
+                }
+                goto unlock;
+            }
+            /* Starts to update timeout on case of succeed */
+            ++stats->num_rate_track_completed;
```

Comment:
unrelated comment about same variable:

we increment num_rate_track_completed here, even if the request never had a timeout set. So if this request was sent before we'd gotten 10 responses, it's timeout was never set, and it shouldn't count towards the failure_rate